apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java

```diff
@@ -1103,50 +1030,163 @@ public Bundle searchByCoverageContractAndYearMonth(
+      LocalDate yearMonth, String contractId, PatientLinkBuilder paging) {
+    // Create the query to run.
+    CriteriaBuilder builder = entityManager.getCriteriaBuilder();
+    CriteriaQuery<String> beneIdCriteria = builder.createQuery(String.class);
+    Root<BeneficiaryMonthly> beneMonthlyRoot = beneIdCriteria.from(BeneficiaryMonthly.class);
+    beneIdCriteria.select(
+        beneMonthlyRoot.get(BeneficiaryMonthly_.parentBeneficiary).get(Beneficiary_.beneficiaryId));
+
+    List<Predicate> wherePredicates = new ArrayList<>();
+    wherePredicates.add(
+        builder.equal(beneMonthlyRoot.get(BeneficiaryMonthly_.yearMonth), yearMonth));
+    wherePredicates.add(
+        builder.equal(beneMonthlyRoot.get(BeneficiaryMonthly_.partDContractNumberId), contractId));
+    if (paging.isPagingRequested() && !paging.isFirstPage()) {
+      wherePredicates.add(
+          builder.greaterThan(
+              beneMonthlyRoot
+                  .get(BeneficiaryMonthly_.parentBeneficiary)
+                  .get(Beneficiary_.beneficiaryId),
+              paging.getCursor()));
+    }
+    beneIdCriteria.where(
+        builder.and(wherePredicates.toArray(new Predicate[wherePredicates.size()])));
+    beneIdCriteria.orderBy(builder.asc(beneMonthlyRoot.get(BeneficiaryMonthly_.parentBeneficiary)));
+
+    // Run the query and return the results.
+    List<String> matchingBeneIds = null;
+    Long beneHistoryMatchesTimerQueryNanoSeconds = null;
+    Timer.Context beneIdMatchesTimer =
+        metricRegistry
+            .timer(
+                MetricRegistry.name(
+                    getClass().getSimpleName(),
+                    "query",
+                    "bene_ids_by_year_month_part_d_contract_id"))
+            .time();
+    try {
+      matchingBeneIds =
+          entityManager
+              .createQuery(beneIdCriteria)
               .setMaxResults(paging.getPageSize() + 1)
               .getResultList();
-
-      // Fetch the benes using the ids
-      return queryBeneficiariesByIdsWithBeneficiaryMonthlys(ids, requestHeader).getResultList();
-    } else {
-      // Fetch benes and their histories in one query
-      return queryBeneficiariesByPartDContractCodeAndYearMonth(
-              contractCode, yearMonth, paging, requestHeader)
-          .setMaxResults(paging.getPageSize() + 1)
-          .getResultList();
+      return matchingBeneIds;
+    } finally {
+      beneHistoryMatchesTimerQueryNanoSeconds = beneIdMatchesTimer.stop();
+      TransformerUtilsV2.recordQueryInMdc(
+          "bene_ids_by_year_month_part_d_contract_id",
+          beneHistoryMatchesTimerQueryNanoSeconds,
+          matchingBeneIds == null ? 0 : matchingBeneIds.size());
     }
   }
+
+  /**
+   * Check that the page size is valid
+   *
+   * @param paging to check
+   */
+  public static void checkPageSize(LinkBuilder paging) {
+    if (paging.getPageSize() == 0) throw new InvalidRequestException("A zero count is unsupported");
```

Comment:
couldn't we just combine this exception? i.e., paging.getPageSize() < 1 : "page size less than 1 (one) is not supported"